### PR TITLE
feat(deepspeed): Deepspeed ZeRO Optimizer with MS-AMP support 

### PR DESCRIPTION
**Pull request by wkcn**
**Description**
In this PR, ZeRO Optimizer with MS-AMP support is integarted into `msamp.deepspeed`.

**Major Revision**
-  add deepspeed ZeRO optimizer with MS-AMP support
-  add ut for zero
- Upgrade pip in docker file.

**Test in internal code**
In FP8Train, run the following script and get around 28.3% top-1 accuracy.
```python
python -m torch.distributed.launch --master_port 1253 --nproc_per_node 8 main.py --cfg configs/deit/deit_mini.yaml --data-path ./cifar100 --batch-size 128 --use-deepspeed --output ./cifar_checkpoints/test1 --opts TRAIN.USE_CHECKPOINT False DATA.DATASET CIFAR100 FP8.USE_FP8_LINEAR True TRAIN.ZERO_STAGE 2
```

### CODE REVIEW

File: tests/deepspeed/test_zero.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    def setUp(self):
+        """Hook method for setting up the test fixture before exercising it."""
+        torch.manual_seed(1000)
+        pass
+
+    def tearDown(self):
+        """Hook method for deconstructing the test fixture after testing it."""
+        pass
+
+    def _backward(self, ds_config):
```
[the remark this answers is not in the document]
Comment (tocean) [reply tests/deepspeed/test_zero.py]: Sure. Done.

**Review by tocean — APPROVED** [review]
LGTM.

File: tests/deepspeed/test_zero.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+import unittest
+
+import torch
+import torch.nn as nn
+
+from msamp import deepspeed
+from msamp.common.dtype.dtypes import Dtypes
+from msamp.nn import LinearReplacer
+from msamp.optim import LBAdam
+
```
Comment (guoshzhao) [inline tests/deepspeed/test_zero.py:14]: Remove blank line.

File: msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py [hunk 3]
```diff
@@ -0,0 +1,992 @@
...
+import torch
+from torch._utils import _flatten_dense_tensors, _unflatten_dense_tensors
+from deepspeed import comm as dist
+
+from msamp.common.tensor import ScalingTensor, ScalingMeta
+from msamp.common.dtype import Dtypes
+from msamp.nn import model_state
+from msamp.common.utils import TransformerEngineWrapper
+
+from deepspeed.runtime.zero.stage_1_and_2 import all_gather_dp_groups, DeepSpeedZeroOptimizer, \
```
Comment (guoshzhao) [inline msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py:18]: Better put deepspeed import together?

File: msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py [hunk 4]
```diff
@@ -0,0 +1,992 @@
...
+    get_accelerator, move_to_cpu, logger, see_memory_usage
+
+SINGLE_PARTITION_OF_FP8_GROUPS = 'single_partition_of_fp8_groups'
+
+MASTER_WEIGHT_QTYPE = Dtypes.kfloat16
+WEIGHT_GRAD_QTYPE = Dtypes.kfloat8_e4m3
+WEIGHT_QTYPE = Dtypes.kfloat8_e4m3
```
Comment (guoshzhao) [inline msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py:25]: Will not change the weight type in the future?

**Review by guoshzhao — APPROVED** [review]

File: tests/deepspeed/test_zero.py (hunk 2 above)
Comment (wkcn) [reply tests/deepspeed/test_zero.py]: Done. Thanks!

File: msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py (hunk 3 above)
Comment (wkcn) [reply msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py]: Done.

File: msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py (hunk 4 above)
Comment (wkcn) [reply msamp/deepspeed/runtime/zero/fp8_stage_1_and_2.py:23]: Yes. We will provide an API or new configuration to change them.
